data/src/test/java/org/apache/iceberg/TestScanPlanningAndReporting.java

```diff
@@ -0,0 +1,210 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg;
+
+import static org.apache.iceberg.Files.localInput;
+import static org.apache.iceberg.types.Types.NestedField.required;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.Assert.assertTrue;
+
+import java.io.File;
+import java.io.IOException;
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.List;
+import org.apache.iceberg.data.GenericRecord;
+import org.apache.iceberg.data.parquet.GenericParquetWriter;
+import org.apache.iceberg.expressions.Expressions;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.io.FileAppender;
+import org.apache.iceberg.metrics.LoggingScanReporter;
+import org.apache.iceberg.metrics.ScanReport;
+import org.apache.iceberg.metrics.ScanReporter;
+import org.apache.iceberg.parquet.Parquet;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.types.Types;
+import org.junit.Before;
+import org.junit.Test;
+
+public class TestScanPlanningAndReporting extends TableTestBase {
+
+  private final Schema schema =
+      new Schema(
+          required(1, "id", Types.IntegerType.get()), required(2, "x", Types.StringType.get()));
+
+  PartitionSpec partitionSpec = PartitionSpec.builderFor(schema).build();
+
+  private Table table;
+  private TestScanReporter reporter;
+
+  public TestScanPlanningAndReporting() {
+    super(2);
+  }
+
+  @Before
+  @Override
+  public void setupTable() throws Exception {
+    super.setupTable();
+    reporter = new TestScanReporter();
+    table =
+        TestTables.create(
+            tableDir,
+            "scan-planning-x",
+            schema,
+            partitionSpec,
+            SortOrder.unsorted(),
+            formatVersion,
+            reporter);
+    GenericRecord record = GenericRecord.create(schema);
+    record.setField("id", 1);
+    record.setField("x", "23");
+    GenericRecord record2 = GenericRecord.create(schema);
+    record2.setField("id", 2);
+    record2.setField("x", "30");
+    GenericRecord record3 = GenericRecord.create(schema);
+    record3.setField("id", 3);
+    record3.setField("x", "45");
+    GenericRecord record4 = GenericRecord.create(schema);
+    record4.setField("id", 3);
+    record4.setField("x", "51");
+    DataFile dataFile = writeParquetFile(table, Arrays.asList(record, record3));
+    DataFile dataFile2 = writeParquetFile(table, Arrays.asList(record2));
+    DataFile dataFile3 = writeParquetFile(table, Arrays.asList(record4));
+    table.newFastAppend().appendFile(dataFile).appendFile(dataFile2).appendFile(dataFile3).commit();
+    table.refresh();
+  }
+
+  @Test
+  public void testScanPlanningWithReport() throws IOException {
```

Comment:
I think that we need to validate that each counter is being accumulated correctly. That's a lot of work, so we may want to scope this PR down and add just a few high-level counters and then get more specific. That way we can add counters and tests together to make sure we get all the cases.

How about adding just the manifest counters now and we can expand after that? Really whatever subset you have done and want to get tested is a good start. Could also be result counters.

Reply:
that sounds good to me. Let me remove some counters and push a new version